### DISCUSSION TITLE
Standardize `IIF.Mixture`

### DISCUSSION THREAD

**Discussion by dehann** [JuliaRobotics/Caesar.jl#808] [Ideas]
We want to be able to make mixture distributions by combining different componets, for example `MvNormal` with a `ManifoldKernelDensity` with a `FluxDistribution`.  The current code is first implementation which reversed the API, the current older code does:
```julia
IIF.Mixture(Pose2Pose2, (MvNormal, ManifoldKernelDensity, FluxDistr), (0.5,0.2,0.3))
```

However, all throughout stats literature as well as Julia/StatsBase, `Mixture` is used as follows (and what i'm suggesting we do as well):
```julia
Pose2Pose2( IIF.Mixture((MvN, MKD, FluxDistr), (0.5,0.2,0.3)) )
```

Then also the name conflict with StatsBase.  We cannot use `StatsBase.Mixture` because they only allow `<:Distribution`, yet we need to allow more diverse components `<:IIF.SamplableBelief`.  So my suggestion is we just dont export, and write all the documentation as `IIF.Mixture(...)`.

**Comment by dehann** [comment]
Mixture should just be a density/distribution in its own right.  We can sample from it like any other, it's integral still adds up to 1, etc.

**Comment by dehann** [comment]
And also just to link between related information:
https://juliarobotics.org/Caesar.jl/latest/concepts/dataassociation/#Mixture-Models

**Comment by GearsAD** [comment]
Got it. Should I wait for the new Mixturesv2 factors, or can we go ahead for now with the existing pattern and then make a Mixturesv2 factor?

**Reply by dehann** [reply]
Think for time sensitivity on ICRA we've gone with existing pattern for now.  I would like to fix this to the better standard as soon we can though -- before the current 'bad' pattern propagates too far.

Question becomes, what is cleanest way to update the SDKs once we change IIF?

**Comment by Affie** [comment]
Can we have <:IIF.SamplableBelief be `Distributions.Sampleable`

**Reply by dehann** [reply]
Yes, but not directly.  We probably should adopt a blend: 
```julia
IIF.SamplableBelief = Union{<:Distributions.Sampleable, AMP.MKD, IIF.AliasingSS, IIF.HSD, IIF.LSD, ...}
```

https://github.com/JuliaRobotics/IncrementalInference.jl/blob/0f63b500191e35b3ec34f86bbe1185fce756ee80/src/entities/BeliefTypes.jl#L22

**Reply by Affie** [reply]
What I was wondering here is if we can have AMP and IIF inherit from `Distributions.Sampleable` but that is probably too restricting.

This is maybe one of those cases where a `SamplableBelief` trait would have worked nicely. I'll play with that a bit at some stage.

**Comment by Affie** [comment]
Is there any use in having a mixture of factor mechanics? For example a Pose2Pose2 and FluxPose2Pose2 factor. I guess one can just create a custom factor that handles the mixing.

**Reply by dehann** [reply]
There is a big need for having a mixture of Distribution with Flux (or MKD, or AliasingSampler, ...).  This mixture would plug into any factor.

I regret using the word mechanics here :-P

**Comment by Affie** [comment]
Once we standardize, we can remove:
```julia
getFactorMechanics(f::AbstractFactor) = f
getFactorMechanics(f::Mixture) = f.mechanics
```
